pkg/controller/cluster/cluster_reconcile.go

```diff
@@ -75,21 +81,68 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			Kind:    instance.Kind,
 		}
 
+		if instance.Spec.GitRepoTemplate == nil {
+			instance.Spec.GitRepoTemplate = &synv1alpha1.GitRepoTemplate{}
+		}
+
 		if len(instance.Spec.GitRepoTemplate.DisplayName) == 0 {
 			instance.Spec.GitRepoTemplate.DisplayName = instance.Spec.DisplayName
 		}
 
 		instance.Spec.GitRepoTemplate.DeletionPolicy = instance.Spec.DeletionPolicy
 
+		nsName := request.NamespacedName
+		nsName.Name = instance.Spec.TenantRef.Name
+
+		tenant := &synv1alpha1.Tenant{}
+
+		if err := r.client.Get(context.TODO(), nsName, tenant); err != nil {
+			return fmt.Errorf("Couldn't find tenant %w", err)
+		}
+
+		if instance.Spec.Facts == nil {
+			instance.Spec.Facts = &synv1alpha1.Facts{}
+		}
+
+		tmplInput := tmplData{
+			ClusterID: instance.Name,
+			TenantID:  tenant.Name,
+			Facts:     *instance.Spec.Facts,
+		}
+
+		if len(instance.Spec.GitRepoTemplate.RepoName) == 0 {
```

Comment:
This seems to be a recurring pattern. It might be worth to abstract that away into a helper function. This would make the code easier to test (test the helper in isolation). It also would make it easier to read.